File: Find Smallest Letter Greater Than Target.py
# https://leetcode.com/problems/find-smallest-letter-greater-than-target/

from bisect import bisect

class Solution:
    
    def binarySearch(self, letters, target, leftPtr, rightPtr):
            
        # Continue if pointers don't overlap
        if leftPtr <= rightPtr:
            
            # Create middle pointer
            midPtr = (leftPtr + rightPtr) // 2
            
            # If element at middle pointer is less than or equal to target, go to right-half of list
            if letters[midPtr] <= target:
                return self.binarySearch(letters, target, midPtr + 1, rightPtr)
            
            # If element at middle pointer greater than target, go to left-half of list
            elif letters[midPtr] > target:
                return self.binarySearch(letters, target, leftPtr, midPtr - 1)
            
        return leftPtr
        
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        
        # SOLUTION #1
          
        # Get index of where target should be
        idx = self.binarySearch(letters, target, 0, len(letters) - 1)
        
        # Modulo is used in case binarySearch() suggests to insert target after the last element. If that's the case, return 0
        return letters[idx % len(letters)]
        
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        
        # SOLUTION #2
        
        # If list is empty, exit function
        if letters is None:
            return
        
        # Use bisect() to get index of where target should be inserted
        """NOTE: bisect() returns an index + 1 of an element that is smaller than target. The index after the small element features an element that is larger
                    than target, so target will fit right inbetween them"""
        
        idx = bisect(letters, target)
        
        # Modulo is used in case bisect() suggests to insert target after the last element. If that's the case, return 0
        return letters[idx % len(letters)]

File: Find Smallest Letter Greater Than Target.java
// https://leetcode.com/problems/find-smallest-letter-greater-than-target/

class Solution {
    
    public int binarySearch(char[] letters, char target, int leftPtr, int rightPtr) {
        
        // If pointers don't overlap, continue
        if (leftPtr <= rightPtr) {
            
            // Create a middle pointer
            int midPtr = (leftPtr + rightPtr) / 2;
            
            // If element at midPtr is less than or equal to target, update midPointer
            /* NOTE: We're trying to find the index of the SMALLEST, LARGEST element of target.
                I.e., Of all the possible elements that is larger than target, return the smallest of them */
            if (letters[midPtr] <= target) {
                return binarySearch(letters, target, midPtr + 1, rightPtr);
            }
            else {
                return binarySearch(letters, target, leftPtr, midPtr - 1);
            }
        }
        return leftPtr;
    }
    
    public char nextGreatestLetter(char[] letters, char target) {
        
        // If array is empty, exit function
        if (letters == null) {return 'a';}
        
        // Find index of where to insert target
        int targetIdx = binarySearch(letters, target, 0, letters.length - 1);
        return letters[targetIdx % letters.length];
    }
}
